### PR TITLE
Support Scheduling Volunteer Shifts

### PR DESCRIPTION
**Pull request by bcholmes**
This PR implements volunteer shift support to PlanZ. 

I think there's meaningful overlap between the idea of "program scheduling" and "volunteer scheduling": both deal with people, times, etc. At a con, it's often fruitful to consider when someone's volunteer schedule conflicts with their panel schedule. (I also intend to show both types of data in the WisCon mobile app -- a long-standing request is that the mobile app help with reminding folks of their volunteer shifts).

Importantly, this PR introduces two screens that are implemented in React. The React code must be built (see instructions in the client directory) and the resulting .js deployed to your web server.

The functionality is implemented using two additional stages: Volunteer Set-up and Volunteer Sign-up. I haven't bothered with a variable in the db_name file: my thinking is that you can exclude volunteer functionality by simply not turning on those phases.

Volunteer set up involves setting up two pieces of data -- jobs and shifts -- like so:
Screen Shot 2022-04-08 at 5 50 35 PM

During the Volunteer sign-up, members can add a shift to their schedule, like so:

Screen Shot 2022-04-08 at 5 50 52 PM

Screen Shot 2022-04-08 at 5 50 59 PM

Lastly, there's a report of the sign-ups:
Screen Shot 2022-04-08 at 5 54 40 PM

I've build the Rest endpoints using a more OO programming style, per James' discussions in the Discussions area. I've also tried to follow the new DB standards.

### CODE REVIEW

**Comment by lostcarpark** [comment]
This looks like impressive work.

I certainly agree there's a meaningful overlap between programme schedule and volunteer schedule, and I can see the logic in having them integrated. However, my personal preference is for applications to concentrate on doing a limited number of things well, rather than trying to be an all encompassing solution and perhaps end up doing a lot of things in mediocre ways.

I think there's massive value in the volunteer system having access to the programme data, and probably vice versa. I'd just like conventions to be able to use either system without having to also take the other (I know they can install Zambia/PlanZ and just use the bits they need).

I think this is a very good example of the benefits of building a modular system, that allows additional modules to be installed as required. It would be ideal if the volunteer "module" could be only enabled if needed, or was even packaged as a separate module. Even better if we broke down PlanZ so that the core was just a scaffolding that other components sat on, so conventions could potentially use the volunteer management functions even if they were using a different system for their programme (though they wouldn't get the benefits of finding conflicts between the schedules).

Anyway, I'm not suggesting we don't add volunteer scheduling, but just try to keep things so that they could be picked apart later. That should be relatively easy since the front end is a React app, providing the React components for volunteers are kept separate from any React work done for the Programme side of PlanZ. So if we add React elements to other parts of PlanZ, I'd encourage them to use a separate React app (or apps), to facilitate future modularisation.

I don't want to sound like I'm not in favour of this work.

I'm at Eurocon in Luxembourg this weekend, and I have Eastercon next weekend. I'm looking forward to looking at this, but I'm not sure I'll get time until after Eastercon.

**Comment by bcholmes** [comment]
I confess that I like Blender's approach: they bundle a lot of things as "extensions", many of which are created by the Blender developers and are pre-bundled with the app although not initially enabled, and many of which are downloadable as third-party add-ons (free or otherwise). 

I certainly like the idea of modularity; I would just also like the process of making things available to be simple. (Again, I think WordPress does a really good job of this: it's easy to search through a registry of plugins, and easy to install them).

The parts that are tricky have to do with things like "how do we add menu items to the nav bar" and "what code in the base can be considered an API"? These are solvable problems, but require work to get there.

File: webpages/api/volunteer/volunteer_job_model.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                mysqli_stmt_close($stmt);
+            } else {
+                throw new DatabaseSqlException("Query could not be executed: $query");
+            }
+        } else {
+            $query = <<<EOD
+            UPDATE volunteer_job
+               SET job_name = ?, is_online = ?, job_description = ?
+             WHERE id = ?;
+            EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_job_model.php:84]: Another EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                J.is_online,
+                J.job_description,
+                sum((case when PHVS.badgeid is not null then 1 else 0 end)) as signup_count
+             FROM volunteer_shift S
+             JOIN volunteer_job J ON (J.id = S.volunteer_job_id)
+        LEFT JOIN participant_has_volunteer_shift PHVS ON (PHVS.volunteer_shift_id = S.id)
+           GROUP BY S.id, S.min_volunteer_count, S.max_volunteer_count, S.from_time,
+                    S.to_time, S.location, J.id, J.job_name, J.is_online, J.job_description
+           ORDER BY S.from_time, J.job_name;
+        EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:44]: Another EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                sum((case when PHVS.badgeid is not null then 1 else 0 end)) as signup_count
+            FROM
+                volunteer_shift S
+            JOIN volunteer_job J ON (J.id = S.volunteer_job_id)
+        LEFT JOIN participant_has_volunteer_shift PHVS ON (PHVS.volunteer_shift_id = S.id)
+        WHERE S.id in (select volunteer_shift_id from participant_has_volunteer_shift where badgeid = ?)
+        GROUP BY S.id, S.min_volunteer_count, S.max_volunteer_count, S.from_time,
+                    S.to_time, S.location, J.id, J.job_name, J.is_online, J.job_description
+           ORDER BY S.from_time, J.job_name;
+        EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:78]: EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    public static function persist($db, $volunteerShift) {
+        $fromTime = $volunteerShift->fromTime->format("Y-m-d H:i:s");
+        $toTime = $volunteerShift->toTime->format("Y-m-d H:i:s");
+
+        if ($volunteerShift->id == null) {
+            $query = <<<EOD
+            INSERT INTO volunteer_shift
+                    (volunteer_job_id, from_time, to_time, location, min_volunteer_count, max_volunteer_count)
+            VALUES (?, ?, ?, ?, ?, ?);
+            EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:132]: EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+                mysqli_stmt_close($stmt);
+            } else {
+                throw new DatabaseSqlException("Insert could not be executed: $query");
+            }
+        } else {
+            $query = <<<EOD
+            UPDATE volunteer_shift
+               SET volunteer_job_id = ?, from_time = ?, to_time = ?, location = ?, min_volunteer_count = ?, max_volunteer_count = ?
+             WHERE id = ?;
+            EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:147]: EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    public static function exists($db, $shiftId) {
+        $query = <<<EOD
+        SELECT
+                count(*) as c
+            FROM
+                volunteer_shift S 
+           WHERE id = ?;
+        EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:167]: EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        } else {
+            throw new DatabaseSqlException("Select count(*) command could not be executed: $query");
+        }
+    }
+
+
+    public static function deleteAssignment($db, $badgeId, $shiftId) {
+        $query = <<<EOD
+        DELETE FROM participant_has_volunteer_shift WHERE badgeid = ? and volunteer_shift_id = ?;
+        EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:188]: EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            mysqli_stmt_close($stmt);
+        } else {
+            throw new DatabaseSqlException("Delete command could not be executed: $query");
+        }
+    }
+
+    public static function deleteShift($db, $shiftId) {
+        $query = <<<EOD
+        DELETE FROM volunteer_shift WHERE id = ?;
+        EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:202]: EOD.

File: webpages/api/volunteer/volunteer_shift_model.php [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+            mysqli_stmt_close($stmt);
+        } else {
+            throw new DatabaseSqlException("Delete command could not be executed: $query");
+        }
+    }
+
+    public static function createAssignment($db, $badgeId, $shiftId) {
+        $query = <<<EOD
+        INSERT INTO participant_has_volunteer_shift (badgeid, volunteer_shift_id) values (?, ?);
+        EOD;
```
Comment (LVerhulst4321) [inline webpages/api/volunteer/volunteer_shift_model.php:216]: EOD.

**Comment by LVerhulst4321** [comment]
This is a really nice set of code to include. I like that you used the phases to turn it on and off as I know that not every con would use it. I found a few items to fix - all related to the EOD. Some versions of PHP require that to have no whitespace to the left of it.

I also agree with the concept of modules for various items and tools. There is a lot of work to be done to get the existing code changed to be more amenable (is that the right word?) to that. Little by little though, we can get there.

**Review by LVerhulst4321 — CHANGES_REQUESTED** [review]

File: webpages/api/volunteer/volunteer_shift_model.php (hunk 9 above)
Comment (bcholmes) [reply webpages/api/volunteer/volunteer_shift_model.php]: Resolved in 5640d804b653c68d6a5e0897584e38cd4de4808f

**Comment by bcholmes** [comment]
Commit 5640d804b653c68d6a5e0897584e38cd4de4808f should address the changes requested.

**Comment by LVerhulst4321** [comment]
Looks good!

**Comment by lostcarpark** [comment]
Just out of interest, how do you set up the test React server?

**Comment by bcholmes** [comment]
> Just out of interest, how do you set up the test React server?

Awkwardly, I didn't set up a test React server. I just repeatedly built the code and pushed it to my test Apache server. 

The brainstorm React code uses webpack to have a watching server, but that webpack configuration ends up modifying the index.html in a way that I was trying to avoid.

**Comment by lostcarpark** [comment]
Ooh that sounds a bit tedious. I was thinking of putting the base address of the API endpoints in a variable, which could be blank in production, but could be set to the PlanZ base URL in development.

**Pull request merged**